src/cpu/x64/jit_brgemm_weights_decompression_kernel.cpp

```diff
@@ -61,38 +61,89 @@ void jit_brgemm_weights_decompression_kernel_t<isa>::generate() {
         uni_vmovups(vmm_mask(1), ptr[reg_tmp]);
     }
 
+    auto vmm_lookup = vmm_tmp(0);
+    auto vmm_lookup_low = vmm_tmp(0);
+    auto vmm_lookup_high = vmm_tmp(1);
+    auto vmm_mask8 = vmm_tmp(2);
+    auto vmm_mask7 = vmm_tmp(3);
+    if (jcp_.weights_dt == data_type::nf4) {
+        static const float lookup[16] = {
+            -1.0,
+            -0.6961928009986877,
+            -0.5250730514526367,
+            -0.39491748809814453,
+            -0.28444138169288635,
+            -0.18477343022823334,
+            -0.09105003625154495,
+            0.0,
+            0.07958029955625534,
+            0.16093020141124725,
+            0.24611230194568634,
+            0.33791524171829224,
+            0.44070982933044434,
+            0.5626170039176941,
+            0.7229568362236023,
+            1.0
+        };
+
+        static const int32_t mask8[16] = {
+            8, 8, 8, 8, 8, 8, 8, 8
+        };
+        static const int32_t mask7[16] = {
+            7, 7, 7, 7, 7, 7, 7, 7
+        };
+
+        if (isa == avx2) {
+            mov(reg_tmp, (size_t)lookup);
+            uni_vmovups(vmm_lookup_low, ptr[reg_tmp]);
+            uni_vmovups(vmm_lookup_high, ptr[reg_tmp + 8 * sizeof(float)]);
+            mov(reg_tmp, (size_t)mask8);
+            uni_vmovups(vmm_mask8, ptr[reg_tmp]);
+            mov(reg_tmp, (size_t)mask7);
+            uni_vmovups(vmm_mask7, ptr[reg_tmp]);
+        } else {
+            mov(reg_tmp, (size_t)lookup);
+            uni_vmovups(vmm_lookup, ptr[reg_tmp]);
+        }
+    }
+
     size_t oc_blocks_num = div_up(jcp_.oc_size, vec_size);
     for (size_t ocb = 0; ocb < oc_blocks_num; ocb++) {
+        auto ic_iters = jcp_.decomp_buffer_dt == data_type::bf16 ? jcp_.ic_internal_size : 1;
         if (jcp_.with_scales) {
             if (jcp_.broadcast_scales) {
-                for (size_t ic = 0; ic < jcp_.ic_internal_size; ic++) {
+                for (size_t ic = 0; ic < ic_iters; ic++) {
                     uni_vbroadcastss(vmm_scales(ocb, ic), ptr[reg_scales]);
                 }
             } else {
-                if (jcp_.ic_internal_size > 1) {
-                    uni_vmovups(vmm_tmp(), ptr[reg_scales + ocb * vec_size * sizeof(float)]);
-                    for (size_t ic = 0; ic < jcp_.ic_internal_size; ic++) {
-                        vpermd(vmm_scales(ocb, ic), vmm_mask(ic), vmm_tmp());
+                if (ic_iters > 1 && jcp_.decomp_buffer_dt == data_type::bf16) {
+                    uni_vmovups(vmm_tmp(0), ptr[reg_scales + ocb * vec_size * sizeof(float)]);
+                    for (size_t ic = 0; ic < ic_iters; ic++) {
+                        vpermd(vmm_scales(ocb, ic), vmm_mask(ic), vmm_tmp(0));
                     }
                 } else {
-                    uni_vmovups(vmm_scales(ocb, 0), ptr[reg_scales + ocb * vec_size * sizeof(float)]);
+                    for (size_t ic = 0; ic < ic_iters; ic++) {
```

Comment:
when "if (ic_iters > 1 && jcp_.decomp_buffer_dt == data_type::bf16)" is not true. ic_iters is always 1. Should we just use the original line?

Reply:
Agree. Removed loop